File: sdk-js/src/providers/openai.ts
// Copyright 2026 Harsha Krishne Gowda
// SPDX-License-Identifier: Apache-2.0

/** OpenAI provider integration — wraps the OpenAI client to track usage. */

import type { UsageEvent, TokenBudgetOptions } from '../types.js';
import { resolveConfig } from '../config.js';
import { calculateCost } from '../pricing.js';
import { getCurrentTags } from '../context.js';
import { EventTransport } from '../transport.js';

// Shared transport instance
let _transport: EventTransport | null = null;

/**
 * Wrap an OpenAI client to automatically track token usage.
 *
 * Patches `client.chat.completions.create` in-place and returns the same client.
 *
 * @param client - An OpenAI client instance
 * @param options - TokenBudget configuration options
 * @returns The same client, now patched to track usage
 */
export function wrapOpenAI<T extends Record<string, any>>(
  client: T,
  options: TokenBudgetOptions = {},
): T {
  const config = resolveConfig(options);

  if (!config.enabled) return client;

  if (!_transport) {
    _transport = new EventTransport(config);
  }

  const transport = _transport;
  const chat = client.chat as Record<string, any>;
  const completions = chat?.completions as Record<string, any>;

  if (!completions?.create) {
    throw new Error(
      'TokenBudget: client.chat.completions.create not found. Pass an OpenAI client instance.',
    );
  }

  const originalCreate = completions.create.bind(completions);

  completions.create = async function patchedCreate(
    ...args: any[]
  ): Promise<any> {
    const start = performance.now();
    const response = await originalCreate(...args);
    const elapsedMs = Math.round(performance.now() - start);

    try {
      const model: string = response.model ?? 'unknown';
      const usage = response.usage;
      const inputTokens: number = usage?.prompt_tokens ?? 0;
      const outputTokens: number = usage?.completion_tokens ?? 0;
      const totalTokens = inputTokens + outputTokens;
      const costUsd = calculateCost(model, inputTokens, outputTokens);
      const tags = getCurrentTags();

      const event: UsageEvent = {
        provider: 'openai',
        model,
        input_tokens: inputTokens,
        output_tokens: outputTokens,
        total_tokens: totalTokens,
        cost_usd: costUsd,
        latency_ms: elapsedMs,
        tags,
        metadata: {},
        timestamp: Date.now() / 1000,
      };

      transport.send(event);
    } catch {
      // Never let tracking errors affect the user
    }

    return response;
  };

  return client;
}

/** Flush and shut down the OpenAI transport. Exported for testing. */
export async function _shutdownOpenAI(): Promise<void> {
  if (_transport) {
    await _transport.shutdown();
    _transport = null;
  }
}

/** Reset the shared transport. Exported for testing. */
export function _resetOpenAI(): void {
  _transport = null;
}

File: sdk/tokenbudget/providers/openai.py
# Copyright 2026 Harsha Krishne Gowda
# SPDX-License-Identifier: Apache-2.0

"""OpenAI provider integration."""
from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING, Any

from tokenbudget.providers.base import BaseProvider
from tokenbudget.pricing import calculate_cost
from tokenbudget.types import UsageEvent

if TYPE_CHECKING:
    from tokenbudget.transport import EventTransport

logger = logging.getLogger(__name__)


class OpenAIProvider(BaseProvider):
    """Provider integration for the OpenAI Python client."""

    def detect(self, client: Any) -> bool:
        """Return True if the client is from the openai package."""
        return type(client).__module__.startswith("openai")

    def extract_event(self, response: Any, latency_ms: int) -> UsageEvent:
        """Extract a UsageEvent from an OpenAI chat completion response."""
        from tokenbudget.context import get_current_tags

        model = getattr(response, "model", "unknown")
        usage = getattr(response, "usage", None)

        input_tokens = getattr(usage, "prompt_tokens", 0) if usage else 0
        output_tokens = getattr(usage, "completion_tokens", 0) if usage else 0
        total_tokens = input_tokens + output_tokens

        cost = calculate_cost(model, input_tokens, output_tokens)
        tags = get_current_tags()

        return UsageEvent(
            provider="openai",
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=total_tokens,
            cost_usd=cost,
            latency_ms=latency_ms,
            tags=tags,
        )

    def patch(self, client: Any, transport: "EventTransport") -> None:
        """Monkey-patch client.chat.completions.create to track usage."""
        original_create = client.chat.completions.create

        def patched_create(*args: Any, **kwargs: Any) -> Any:
            start = time.perf_counter()
            response = original_create(*args, **kwargs)
            elapsed_ms = int((time.perf_counter() - start) * 1000)
            try:
                event = self.extract_event(response, latency_ms=elapsed_ms)
                transport.send(event)
            except Exception as exc:
                logger.warning("TokenBudget: failed to track OpenAI event: %s", exc)
            return response

        client.chat.completions.create = patched_create
